drivers/can/sja1000.h

```diff
@@ -0,0 +1,911 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __DRIVERS_CAN_SJA1000_H
+#define __DRIVERS_CAN_SJA1000_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Default interrupts - Exclude data overrun (bit[3]) and brp_div (bit[4])
+ */
+
+#define SJA1000_DEFAULT_INTERRUPTS 0xe7
+
+/* Number of standard/extended filters */
+
+#define SJA1000_NFILTERS           1
+
+/* SJA1000_MODE_REG register
+ * Mode Register
+ */
+
+#define SJA1000_MODE_REG(baseaddr) (baseaddr + 0x0)
+
+/* SJA1000_RX_FILTER_MODE : R/W; bitpos: [3]; default: 0;
+ * This bit is used to configure the filter mode. 0: Dual filter mode; 1:
+ * Single filter mode.
+ */
+
+#define SJA1000_RX_FILTER_MODE     (BIT(3))
+#define SJA1000_RX_FILTER_MODE_M                                           \
+  (SJA1000_RX_FILTER_MODE_V << SJA1000_RX_FILTER_MODE_S)
+#define SJA1000_RX_FILTER_MODE_V 0x00000001
+#define SJA1000_RX_FILTER_MODE_S 3
+
+/* SJA1000_SELF_TEST_MODE : R/W; bitpos: [2]; default: 0;
+ * 1: Self test mode. In this mode the TX nodes can perform a successful
+ * transmission without receiving the acknowledge signal. This mode is often
+ * used to test a single node with the self reception request command.
+ */
+
+#define SJA1000_SELF_TEST_MODE   (BIT(2))
+#define SJA1000_SELF_TEST_MODE_M                                           \
+  (SJA1000_SELF_TEST_MODE_V << SJA1000_SELF_TEST_MODE_S)
+#define SJA1000_SELF_TEST_MODE_V 0x00000001
+#define SJA1000_SELF_TEST_MODE_S 2
+
+/* SJA1000_LISTEN_ONLY_MODE : R/W; bitpos: [1]; default: 0;
+ * 1: Listen only mode. In this mode the nodes will only receive messages
+ * from the bus, without generating the acknowledge signal nor updating the
+ * RX error counter.
+ */
+
+#define SJA1000_LISTEN_ONLY_MODE (BIT(1))
+#define SJA1000_LISTEN_ONLY_MODE_M                                         \
+  (SJA1000_LISTEN_ONLY_MODE_V << SJA1000_LISTEN_ONLY_MODE_S)
+#define SJA1000_LISTEN_ONLY_MODE_V 0x00000001
+#define SJA1000_LISTEN_ONLY_MODE_S 1
+
+/* SJA1000_RESET_MODE : R/W; bitpos: [0]; default: 1;
+ * This bit is used to configure the operating mode of the SJA1000
+ * Controller. 1: Reset mode; 0: Operating mode.
+ */
+
+#define SJA1000_RESET_MODE         (BIT(0))
+#define SJA1000_RESET_MODE_M       (SJA1000_RESET_MODE_V << SJA1000_RESET_MODE_S)
+#define SJA1000_RESET_MODE_V       0x00000001
+#define SJA1000_RESET_MODE_S       0
+
+/* SJA1000_CMD_REG register
+ * Command Register
+ */
+
+#define SJA1000_CMD_REG(baseaddr)  (baseaddr + 0x4)
+
```

Comment:
I would not use original SJA1000 offsets multiplied by four when the implementation should be generic for all SJA1000 based cards. Many of them maps registers with span 1. So I would use
```
#define SJA1000_MODE_REG_OFFS  0x00
#define SJA1000_CMD_REG_OFFS    0x01
#define SJA1000_STATUS_REG    0x02
```
and then provide shift left by two and base addition in putreg and getreg specific for given chip mapping. There is some problem with code overhead which I suggest to resolve for really small chips by possibility to provide define controlled option to replace indirect putreg and getreg by direct inline calls with with inline single chip specific versions of putreg and getreg, so versions would be

- generic - with indirect calls when span4 or span1 version of putreg can be specified in the private structure
- span4 - with fixed 4 span 
- span1 - with fixed 1 span